On Chain Circuit Breaker

bayological opened the pull request:
### Description

This change will add an automated on-chain circuit breaker to limit trading via mento when prices shift more than a calculated threshold. Exchanges must manually be added to the BreakerBox to enable checking of price change. 

- Added MedianDeltaBreaker contract to detect shift gt a calculated amount in oracle reports
- Added BreakerBox contract to initiate checks, store exchange trading modes & references to breakers
- Added modifier in Exchange.sol to check trading mode for an exchange when a buy/sell is initiated
- Added variable to store last median in Sorted oracles when current median changes

More info [here](https://forum.celo.org/t/draft-proposal-on-chain-circuit-breaker/3786)

### Other changes

N/A

### Tested

- Unit tests
- Integration tests

### Related issues

N/A

### Backwards compatibility
Remove exchange from BreakerBox to stop checks. 

 
### Documentation

In Progress

bayological committed:
feat(medianDeltaBreaker): Add contract & store last median in sorted oracles

bayological committed:
feat(medianDeltaBreaker): add events, threshold variables & setters

bayological committed:
feat(medianDeltaBreaker): add breaker box interface

bayological committed:
feat(medianDeltaBreaker): add BreakerBox contract & owner functions

bayological committed:
Merge branch 'master' into bayological/onchain-circuit-breaker

bayological committed:
feat(medianDeltaBreaker): remove redundant list checks & add constructor

bayological committed:
test(priceDeviationFilter): add empty test contract for breakerbox

bayological committed:
test(medianDeltaBreaker): switch to floating pragma :(

bayological committed:
feat(medianDeltaBreaker): add breaker checks & views

bayological committed:
test: add breaker & init tests

bayological committed:
feat(BreakerBox): add checkBreakers implementation

bayological committed:
feat(BreakerBox): Add last updated block &
manual trading mode update

bayological committed:
test(BreakerBox): add tests for restricted
breaker functions

bayological committed:
feat(BreakerBox): remove unnecessary zero address check

bayological committed:
test(BreakerBox): add tests for restricted exchange functions

bayological committed:
feat(BreakerBox) : add validation for exchange functions

bayological committed:
test(BreakerBox): additional addExchange tests

bayological committed:
test(BreakerBox): remove breaker should reset in use trading modes

bayological committed:
test(BreakerBox): check exchange is popped on remove

bayological committed:
feat: add breakerbox call to exchange

bayological committed:
test: update exchange tests

bayological committed:
feat(BreakerBox): log reset attempts

bayological committed:
test(BreakerBox): add checkBreaker tests

bayological committed:
feat: remove breaker box from exchange initilizer

bayological committed:
test(medianDeltaBreaker): add test contract

bayological committed:
feat(medianDeltaBreaker): add logic to check median change

bayological committed:
refactor: cleanup variables

bayological committed:
test(BreakerBox): keep test names consistent

bayological committed:
refactor: add modifier for duplicate check

bayological committed:
test(MedianDeltaBreaker): test restricted funcs

bayological committed:
test(MedianDeltaBreaker): add shouldTrigger tests

bayological committed:
feat(BreakerBox): Update tryReset condition

bayological committed:
feat: only store trading mode in breakerbox

bayological committed:
test: update calls to breaker.getTradingMode

bayological committed:
feat: remove requirement of exchange being added

bayological committed:
Merge branch 'master' into bayological/onchain-circuit-breaker

bayological committed:
feat: use simple threshold for breaker

bayological committed:
style: lint fix

bayological committed:
test(Exchange): set breakerBox address & update test

bayological committed:
refactor: remove breakerBox from initilizer

soloseng commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,300 @@
+ * @title   BreakerBox
+ * @notice  The BreakerBox stores references to Mento exchanges and breakers.
+ */
+contract BreakerBox is IBreakerBox, Initializable, UsingRegistry {
```
nit: using solidity style guide for function ordering: https://docs.soliditylang.org/en/v0.5.13/style-guide.html#order-of-functions

bayological replied:
I'm against following that order. I think following this strictly actually makes the code harder to read. The current order groups the admin-only getters & setters then the external functions, which is slightly easier to follow.

bayological committed:
style: update comments

bayological committed:
refactor: reduce storage of tradingModeInfo

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+ *          to determine whether or not buying or selling should be allowed for a
+ *          specified exchange. The contract stores references to all breakers
+ *          that hold criteria to be checked, exchanges that
+ *          can make use of the BreakerBox & their current trading
```
period at the end

m-chrzan commented on packages/protocol/contracts/stability/interfaces/IBreakerBox.sol:
```diff
@@ -0,0 +1,100 @@
+   *      LastUpdatedTime helps to check cooldown.
+   *      LastUpdatedBlock helps to determine if check should be executed.
+   */
+  struct TradingModeInfo {
```
We've usually put all struct definitions in the actual contract, not the interface

bayological replied:
Fair enough. Will be keeping this one here though, I feel the separation makes the BreakerBox contract less busy.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+  /* ==================== State Variables ==================== */
+
+  address[] public exchanges;
+  // Maps exchange address to its current trading mode info
```
Period at the end, here and throughout solidity comments.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+  // Maps exchange address to its current trading mode info
+  mapping(address => TradingModeInfo) public exchangeTradingModes;
+  mapping(uint64 => address) public tradingModeBreaker;
+  mapping(address => uint64) public breakerTradingMode;
```
Not very clear what these are for, add descriptive comments

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+    require(!isBreaker(breaker), "This breaker has already been added");
+    require(
+      tradingModeBreaker[tradingMode] == address(0),
+      "There is already a breaker added with the same trading mode"
```
Do we really have only one breaker per trading mode?

bayological replied:
Yes, that's right. This will help to simplify the logic and the code. One breaker could have multiple conditions if necessary.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+  // Ordered list of breakers to be checked.
+  LinkedList.List private breakers;
+
+  modifier validateBreaker(address breaker, uint64 tradingMode) {
```
A few nitty notes:
* Modifier names are usually non-verbs, so maybe something like `validBreaker`?
* This seems to specifically validate that this is valid info for a *new* breaker, so maybe `validNewBreaker`?
* Finally, if this is used for only one function, would it be cleaner to just inline in that function? Personally, never was a big fan of relying on modifiers for all sorts of checks.

bayological replied:
How about isValidNewBreaker?

The modifier is used on addBreaker & insertBreaker. I agree with it being cleaner to keep it inline, the modifier was to reduce duplication.

bayological replied:
renamed to onlyValidBreaker

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @param breaker The address of a breaker to be added
```
Feels a little weird that there's an arbitrary choice of needing to initialize with exactly one breaker. Why not allow an array to be passed in, or alternatively not set any in initializer, depend on breakers being added manually afterwards?

bayological replied:
Agree with you on this. Initially had an array then thought it would be unnecessary code as we only have one breaker. Will remove from the initializer completely.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+  /**
+   * @param breaker The address of a breaker to be added
+   * @param tradingMode The trading mode of the breaker to be added
+   * @param _exchanges Exchanges to be added
```
So sounds like each breaker is always applied to each exchange, is that desired? I guess if the breaker's logic itself can differentiate between exchanges that should be fine, but just want to double check this is the desired design.

bayological replied:
That's right. A breaker isn't specific to a particular exchange.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+   * @param prevBreaker The address of the breaker that should come before the new breaker.
+   * @param nextBreaker The address of the breaker that should come after the new breaker.
+   */
+  function insertBreaker(
```
Why is this necessary over `addBreaker`, does ordering of breakers affect anything? Also, would OpenZeppelin's `EnumberableSet` be a better struct to hold these breakers?

bayological replied:
The order of the breakers is important.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+    TradingModeInfo memory tradingModeInfo;
+
+    for (uint256 i = 0; i < activeExchanges.length; ++i) {
+      tradingModeInfo = exchangeTradingModes[activeExchanges[i]];
```
Is the Solidity compiler smart enough to not read the whole struct into memory? Or should the above `tradingModeInfo` be `storage` instead?

m-chrzan replied:
Ah, looking at the struct, looks like right now it fits in a single storage slot anyways. However, should it ever expand, this could become less efficient (and it'd be hard to remember to come back here to change it).

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+      "Exchange is not a reserve spender"
+    );
+
+    info.tradingMode = 0; // Default trading mode (Bi-directional).
```
Just want to double check this is correct. In the forum post you said

> Initially, there will be two trading modes, bi-directional mode(1) & no trading(0)

this line implies the numbering is the other way around.

bayological replied:
I switched it around for simplicity. Will update the post.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+   */
+  function removeExchange(address exchange) external onlyOwner {
+    uint256 exchangeIndex;
+    for (uint256 i = 0; i < exchanges.length; ++i) {
```
Nit: `i++` for consistency with other loops.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+    address[] memory activeExchanges = exchanges;
+    TradingModeInfo memory tradingModeInfo;
+
+    for (uint256 i = 0; i < activeExchanges.length; ++i) {
```
Nit: i++ for consistency with other loops throughout the codebase.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+  function removeExchange(address exchange) external onlyOwner {
+    uint256 exchangeIndex;
+    for (uint256 i = 0; i < exchanges.length; ++i) {
+      if (exchanges[i] == exchange) {
```
Would using OpenZeppelin's EnumerableSet be more efficient, allow us to simplify the logic here?

bayological replied:
I'm not sure about that but will check.
The [version](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/58a3368215581509d05bd3ec4d53cd381c9bb40e/contracts/utils/EnumerableSet.sol) we're tied to will do some additional storage writes to shift the index of the address being deleted. The EnumerableSet.add function will also increase the cost of our addExchange function.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+    // if it doesn't we can assume this exchange has not been added.
+    if (info.lastUpdatedTime == 0) {
+      return 0;
+    }
```
Or should this be a require? No way to distinguish between default trading mode and unregistered exchange.

bayological replied:
An unregistered exchange will always have a last updated time == 0 but an exchange in the default trading mode will always have last updated > 0. 
The intention here is to allow trades from an unregistered exchange. Initially this was a require but it feels like better UX to allow trades instead of reverting. This makes it possible to have the breakerBox deployed and only active for certain exchanges.  WDYT?

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+      return 0;
+    }
+
+    // Check if a breaker has non default trading mode & try to reset
```
Wording nit: would better wording be "and reset if we should"?

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+
+      if (tryReset) {
+        bool canReset = breaker.shouldReset(exchangeAddress);
+        if (canReset) {
```
Any reason to separate this into two lines instead of performing the check in the `if`?

m-chrzan replied:
Another thing to consider: are the `ResetAttemptCriteriaFail` and `ResetAttemptNotCool` particularly useful? If not, we could simplify this to a single `if` instead of nesting.

bayological replied:
> Any reason to separate this into two lines instead of performing the check in the `if`?

Nope, they can be inlined.

> Another thing to consider: are the `ResetAttemptCriteriaFail` and `ResetAttemptNotCool` particularly useful? If not, we could simplify this to a single `if` instead of nesting.

They'll be useful for monitoring. If a breaker is tripped these logs will let us easily determine why the trading mode has not been reset yet. Also at some point in the future, I'd like to put together a UI or a dashboard to show this info.

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,305 @@
+        info.lastUpdatedBlock = uint128(block.number);
+        exchangeTradingModes[exchangeAddress] = info;
+        emit BreakerTripped(address(breaker), exchangeAddress);
+        return info.tradingMode;
```
What should happen if multiple breakers should fire at once? Currently we'll just go with the first one that triggers.

bayological replied:
That is by design. The breakers will be ordered by severity/priority.

m-chrzan commented on packages/protocol/contracts/stability/Exchange.sol:
```diff
@@ -67,6 +70,19 @@ contract Exchange is
+  modifier checkTradingMode() {
+    if (address(breakerBox) != address(0)) {
+      require(
+        breakerBox.checkBreakers(address(this)) == 0,
```
Maybe let's define the different mode numbers as named constants, so we're not using magic numbers throughout the code.

m-chrzan replied:
Also, isn't this a pretty expensive computation we're adding to every trade? I don't think I saw `lastUpdatedTime` used to short circuit if the check was already performed in a block.

bayological replied:
> Maybe let's define the different mode numbers as named constants, so we're not using magic numbers throughout the code.

Added

> Also, isn't this a pretty expensive computation we're adding to every trade? I don't think I saw `lastUpdatedTime` used to short circuit if the check was already performed in a block.

Agreed. This is one of the outstanding things to do. The check will require more than I initially thought but will be updating the implementation to short circuit when possible.

m-chrzan reviewed (commented):
Looks nice. Some comments, so far looked at BreakerBox and Exchange changes.

bayological committed:
refactor: code review updates

- Remove breaker from initilizer
- Inline checkBreaker if conditions
- consistent punctuation marks 🤓
- Describe mappings

bayological committed:
refactor: code review updates

- postfix increment

bayological committed:
refactor: add named constant for trading mode

bayological committed:
test(Exchange): add integration tests

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -0,0 +1,136 @@
+
+  // The amount of time that must pass before the breaker can be reset for an exchange.
+  uint256 public cooldownTime;
+  // The allowed threshold for the median price change. Multiplied by 10^24
```
Wondering about wording here. Should we just say "as a Fixidity fraction" or something rather than "Multiplied by 10^24"? One point of confusion could be someone thinking that the Fixidity number itself is then multiplied by 10^24.

bayological replied:
Will update to include that.

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -0,0 +1,136 @@
+
+  /* ==================== Events ==================== */
+
+  event priceChangeThresholdUpdated(uint256 newPriceChangeThreshold);
```
Capitalize event name

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -0,0 +1,136 @@
+  /**
+   * @notice Sets the cooldownTime to the specified value.
+   * @param _cooldownTime The new cooldownTime value.
+   * @dev Should be set to 0 to force a manual reset.
```
Should this info be included in the comment about the `cooldownTime` storage variable above?

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -0,0 +1,136 @@
+    uint256 maxValue = (lastRate.mul(maxPercent)).div(10**24);
+
+    uint256 minPercent = uint256(1 * 10**24).sub(allowedThreshold);
+    uint256 minValue = (lastRate.mul(minPercent)).div(10**24);
```
Can all this arithmetic be done using FixidityLib operations and constants?

bayological replied:
Yes it can. This is leftover from the initial math that used sqrt which can't be done with FixidityLib.

m-chrzan commented on packages/protocol/contracts/stability/interfaces/IBreaker.sol:
```diff
@@ -0,0 +1,50 @@
+  function getCooldown() external view returns (uint256 cooldown);
+
+  /**
+   * @notice Check if the criteria has been met, by a specified exchange, to trigger the breaker.
```
typo: *have* been met

m-chrzan commented on packages/protocol/contracts/stability/interfaces/IBreaker.sol:
```diff
@@ -0,0 +1,50 @@
+  function shouldTrigger(address exchange) external view returns (bool triggerBreaker);
+
+  /**
+   * @notice Check if the criteria to automatically reset the breaker has been met.
```
ditto

m-chrzan commented on packages/protocol/contracts/stability/interfaces/ISortedOracles.sol:
```diff
@@ -10,4 +12,13 @@ interface ISortedOracles {
   function medianRate(address) external view returns (uint256, uint256);
   function numTimestamps(address) external view returns (uint256);
   function medianTimestamp(address) external view returns (uint256);
+  function lastMedianRate(address) external view returns (uint256);
```
Just small thought on naming, maybe `previousMedianRate` would be clearer?

m-chrzan reviewed (commented):
Haven't looked at tests yet

bayological committed:
refactor: code review updates
- Rename addBreaker modifier
- Revert when exchange hasn't been added
-  Clarify comments

bayological committed:
feat: run breaker checks on oracle reports

bayological committed:
test: update tests

bayological committed:
refactor: updates to fix warnings

bayological committed:
refactor: rename to mocks for consistency

bayological committed:
test(SortedOracles): add test to verify call to box

bayological committed:
test(Integration): add tests to check reset

soloseng commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -0,0 +1,299 @@
+   * @notice Checks breakers for the exchange with the specified id.
+   * @param exchangeRegistryId The registryId of the exchange to run checks for.
+   */
+  function checkBreakers(bytes32 exchangeRegistryId) external {
```
Should this be `checkAndSetBreakers` since its also changing the state of breakers

soloseng commented on packages/protocol/contracts/stability/interfaces/IBreaker.sol:
```diff
@@ -0,0 +1,50 @@
+   * @notice Emitted after a breaker has been reset for an exchange.
+   * @param exchange The address of the exchange.
+   */
+  event BreakerReset(address indexed exchange);
```
unused events. Not sure when these event are intended to be used. Also seems like these events might be handled by the `BreakerBox` as well.

bayological committed:
test(SortedOracles): test when report() updates median

bayological committed:
test(Integration): update exchange tests
Covering the following:
- When trading mode for the exchange is bidirectional(default)
- When breakerbox is not set for the exchange
- When median has moved gt allowed threshold then exchange attempted
- When median has moved gt than threshold and back
- When median has moved back & governance reset breaker

bayological committed:
test: remove unused params

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -117,14 +122,17 @@ contract MedianDeltaBreaker is IBreaker, UsingRegistry {
   {
+    // uint256 maxPercent = FixidityLib.newFixed(1).add(allowedThreshold);
+    // uint256 maxValue = (FixidityLib.newFixed(lastRate).multiply(maxPercent)).fromFixed();
+
```
Remove commented out code

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -22,13 +22,14 @@ contract MedianDeltaBreaker is IBreaker, UsingRegistry {
   /* ==================== State Variables ==================== */
 
   // The amount of time that must pass before the breaker can be reset for an exchange.
+  // Should be set to 0 to force a manual reset.
```
Could potentially want a constant (like `REQUIRE_MANUAL_RESET = 0`) to avoid magic numbers.

bayological replied:
The cooldown time isn't read in this contract, I'm not sure how introducing this constant would help.

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -91,22 +92,26 @@ contract MedianDeltaBreaker is IBreaker, UsingRegistry {
-    if (lastMedian == 0) {
+    uint256 previousMedian = sortedOracles.previousMedianRate(stableToken);
+    if (previousMedian == 0) {
       // Last median will be 0 if this exchange is new and has not had a median update yet.
```
Minor nit:
```suggestion
      // Previous median will be 0 if this exchange is new and has not had at least two median updates yet.
```

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -91,22 +92,26 @@ contract MedianDeltaBreaker is IBreaker, UsingRegistry {
+    triggerBreaker = !isWithinThreshold(
+      previousMedian,
+      currentMedian,
+      priceChangeThreshold.unwrap()
```
Does this need to be `unwrapped`? Can't we pass a FixidityFraction to the function?

bayological replied:
This doesn't even need to be here. Can be read directly in the function.

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -117,14 +122,17 @@ contract MedianDeltaBreaker is IBreaker, UsingRegistry {
+    // uint256 maxPercent = FixidityLib.newFixed(1).add(allowedThreshold);
+    // uint256 maxValue = (FixidityLib.newFixed(lastRate).multiply(maxPercent)).fromFixed();
+
     uint256 maxPercent = uint256(1 * 10**24).add(allowedThreshold);
```
Use FixidityLib here, i.e. grab `fixed1()` instead of `10**24`, pass in a FixidityLib fraction as argument.

bayological replied:
I'm against using Fixidity here. It doesn't provide any benefits over the current implementation and costs more. Will add Fixed1 and unwrap.
![Screenshot from 2022-07-28 11-16-06](https://user-images.githubusercontent.com/6872903/181587739-1d14c85e-b04f-4d15-b5a9-907a5f00a1fa.png)

m-chrzan replied:
The benefit of using Fixidity is it helps keep the code clean, without needing to remember the details of the specific fixed point arithmetic we're using here, makes sure all the arithmetic is being done exactly the same throughout the code base, and updates to the lib get propagated to all uses of it.

The cost difference is pretty big (any idea why, btw? is this just from all the `delegatecall`s to a linked library?) but we need to consider the trade-off with maintainability of the code.

m-chrzan commented on packages/protocol/contracts/stability/SortedOracles.sol:
```diff
@@ -37,6 +39,8 @@ contract SortedOracles is ISortedOracles, ICeloVersionedContract, Ownable, Initi
   uint256 public reportExpirySeconds;
   mapping(address => uint256) public tokenReportExpirySeconds;
 
+  IBreakerBox public breakerBox;
```
Should the BreakerBox be just registered in the Registry? Then we don't need this variable, don't need `setBreakerBox`, just look up in Registry. Look up from there both here and in Exchange, that way if BreakerBox address changes, it automatically updates for all Exchanges and SortedOracles, no need to remember to manually update it via governance proposal for each contract individually.

bayological replied:
I initially didn't add it as I didn't want to introduce that dependency. @bowd also suggested it would be best not to add to the registry.

m-chrzan replied:
This is how we handle inter-core contract dependencies all throughout the protocols. If we break this pattern here, we'll have to remember to manually update this, should the address ever change. This would complicate the release process and create more surface for human error, that the Registry pattern is specifically supposed to help us avoid.

Additionally, registering in the Registry is helpful for some other things, e.g. getting core contract addresses in contractkit.

bayological committed:
Update comment

Co-authored-by: Martin <martin@clabs.co>

soloseng commented on packages/protocol/test-sol/stability/Exchange.t.sol:
```diff
@@ -315,6 +318,18 @@ contract Exchange_initializeAndSetters is ExchangeTest {
+    exchange.setBreakerBox(breakerBox);
+  }
+
+  function test_setBreakerBox_shouldUpdateAndEmit() public {
```
This doesn't seem to check for events emitted

soloseng reviewed (commented):
Really good work! Just a few comments, nothing major.

bayological committed:
refactor: code review updates
- Remove unused events
- Remove comments
- Added check for emit

bayological committed:
refactor: break execution on reset attempt

m-chrzan commented on packages/protocol/contracts/stability/BreakerBox.sol:
```diff
@@ -275,9 +277,11 @@ contract BreakerBox is IBreakerBox, Initializable, UsingRegistry {
           emit ResetSuccessful(exchangeAddress, address(breaker));
         } else {
           emit ResetAttemptCriteriaFail(exchangeAddress, address(breaker));
+          return;
```
I'm not sure I understand the exact desired logic of this functions.
If the first breaker can be reset, we'll set trading mode to bidirectional, then if the second breaker actually should not be reset, we won't do anything about it: emit an event, return, and leave the exchange in mode 0. Is this ok, or for some reason not a possible situation?

bayological replied:
At this point there isn't a concept of first or second breaker, there is only the breaker that was triggered for the given exchange. If we hit this codepath we know a breaker has been triggered because the exchange is not in the default mode. The purpose of this particular block is to try and reset the breaker for this exchange. 

> If the ~~first~~ breaker can be reset, we'll set trading mode to bidirectional

This is correct. But also there is no return so we can evaluate the other breakers in case the exchange is in a state that would trigger a breaker

> then if the ~~second~~ breaker actually should not be reset, we won't do anything about it: emit an event, return, and leave the exchange in mode 0.

This is half true. If the breaker should not be reset we do not do anything, emit and return. But in this codepath it's not possible for the trading mode to be 0(default). [Line 265](https://github.com/celo-org/celo-monorepo/blob/06e2de31bf43dfabce4087a35d3d95d60dc27035/packages/protocol/contracts/stability/BreakerBox.sol#L265) above checks this. 

So the desired logic is:
- Check if a breaker has already been tripped for this exchange & try to reset if it should be
- Check if a breaker should be tripped for this exchange & update its trading mode if necessary

Maybe breaking the function into separate smaller functions may help.

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -120,24 +116,20 @@ contract MedianDeltaBreaker is IBreaker, UsingRegistry {
 
-    uint256 maxPercent = uint256(1 * 10**24).add(allowedThreshold);
-    uint256 maxValue = (lastRate.mul(maxPercent)).div(10**24);
+    uint256 maxPercent = uint256(fixed1).add(allowedThreshold);
```
The typecast shouldn't be necessary, right?

m-chrzan commented on packages/protocol/contracts/stability/MedianDeltaBreaker.sol:
```diff
@@ -120,24 +116,20 @@ contract MedianDeltaBreaker is IBreaker, UsingRegistry {
-    uint256 maxPercent = uint256(1 * 10**24).add(allowedThreshold);
-    uint256 maxValue = (lastRate.mul(maxPercent)).div(10**24);
+    uint256 maxPercent = uint256(fixed1).add(allowedThreshold);
+    uint256 maxValue = (prevRate.mul(maxPercent)).div(10**24);
```
Are we guaranteed this multiplication will not overflow? The Fixidity implementation of multiplication allows for much bigger arguments than doing direct multiplication on the unwrapped values.
